modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C

```diff
@@ -102,38 +130,65 @@ SingleVariableReturnMappingSolution::returnMappingSolve(const Real effective_tri
                                                         Real & scalar,
                                                         const ConsoleStream & console)
 {
-  std::stringstream iter_output;
-  std::stringstream * iter_output_ptr = (_output_iteration_info ? &iter_output : nullptr);
+  // construct the stringstream here only if the debug level is set to ALL
+  std::stringstream * iter_output =
+      (_debug_level == DebugLevel::ALL) ? new std::stringstream : nullptr;
 
   if (!_legacy_return_mapping)
   {
-    if (!internalSolve(effective_trial_stress, scalar, iter_output_ptr))
+    // do the internal solve and capture iteration info during the first round
+    // iff full history output is requested regardless of whether the solve failed or succeeded
+    if (!internalSolve(
+            effective_trial_stress, scalar, _output_iteration_info ? iter_output : nullptr))
     {
-      if (iter_output_ptr)
-        throw MooseException(iter_output_ptr->str());
+      // if full history output is only requested for failed solves we have to repeat
+      // the solve a second time
+      if (_debug_level == DebugLevel::ERROR && _output_iteration_info)
+      {
+        // construct string stream and capture iteration history in the second solve
+        iter_output = new std::stringstream;
+        internalSolve(effective_trial_stress, scalar, iter_output);
+        outputIterationSummary(iter_output, _iteration);
+        throw MooseException(iter_output->str());
+      }
+      else if (_debug_level == DebugLevel::NONE)
+        throw MooseException("");
       else
       {
-        internalSolve(effective_trial_stress, scalar, &iter_output);
-        throw MooseException(iter_output.str());
+        outputIterationSummary(iter_output, _iteration);
+        throw MooseException(iter_output->str());
       }
     }
-    else if (iter_output_ptr)
-      console << iter_output_ptr->str();
+    else if (_debug_level == DebugLevel::ALL)
+    {
+      // the solve did not fail but the user requested debug output anyways
+      outputIterationSummary(iter_output, _iteration);
+      console << iter_output->str();
+    }
   }
   else
   {
-    if (!internalSolveLegacy(effective_trial_stress, scalar, iter_output_ptr))
+    //
+    // DEPRECATED LEGACY SOLVE, please remove
```

Comment:
When do you think this will be removed?

Reply:
@bwspenc should chime in here. I see it being used in 3 modules tests, 7 Bison tests, and 3 Bison assessment cases.

Reply:
I'd love to get rid of it now. We just need to go back and figure out why we needed it on those tests and remove it.

Reply:
Not in the scope of this PR, though